sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -61,95 +173,203 @@ Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+
+template<>
+class command_graph<graph_state::modifiable> {
+public:
+  command_graph(const property_list &propList = {});
+  command_graph<graph_state::executable> finalize(context &syclContext) const;
+
+  node add(const std::vector<node>& dep = {});
+
+  template<typename T>
+  node add(T cgf, const std::vector<node>& dep = {});
+
+  node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& dep = {});
+  node add_free(void *data, const std::vector<node>& dep = {});
+};
+
+template<>
+class command_graph<graph_state::executable> {
+public:
+    command_graph() = delete;
+    void update(const command_graph<graph_state::modifiable> &graph);
+};
+}  // namespace ext::oneapi::experimental
+
+// New methods added to the sycl::queue class
+using namespace ext::oneapi::experimental;
+class queue {
+public:
+  bool begin_recording(command_graph<graph_state::modifiable> &graph);
+  bool end_recording();
+  event submit(command_graph<graph_state::executable> graph);
+};
+}  // namespace sycl
+----
+
+=== Node
+
+Node is a class that encapsulates tasks like SYCL kernel functions or host tasks
+for deferred execution. A graph has to be created first, the structure of a
+graph is defined second by adding nodes and edges.
 
 [source,c++]
 ----
 namespace sycl::ext::oneapi::experimental {
-
-  class node{
-  };
+  class node {};
 }
 ----
 
-== Edge
+=== Edge
 
-A dependency between two nodes representing a happens-before relationship. `sender` and `receiver` may be associated to different graphs.
+A dependency between two nodes representing a happens-before relationship.
+`sender` and `receiver` may be associated to different graphs.
 
 [source,c++]
 ----
 namespace sycl::ext::oneapi::experimental {
-
-	// Adding dependency between two nodes.
-    	void make_edge(node sender, node receiver);
+  void make_edge(node sender, node receiver); // Adds a dependency between two nodes
 }
 ----
 
-== Graph
+Parameters:
 
-Graph is a class that represents a directed acyclic graph of nodes. 
-A graph can have different states, can be nested, can have multiple root nodes that are scheduled for execution first and multiple leaf nodes that are scheduled for execution last. The execution of a graph has been completed when all leaf node tasks have been completed.
-Member functions as listed in Table 3 to 6 can be used to add nodes to a graph.
+* `sender` - Node which will be a dependency of `receiver`.
 
-[source,c++]
-----
-namespace sycl::ext::oneapi::experimental {
+* `receiver` - Node which will be dependent on `sender`.
 
-  enum class graph_state{
-    modifiable,
-    executable
-  };
-
-  template<graph_state State=graph_state::modifiable>
-  class command_graph {
-  public:
-    operator command_graph<graph_state::executable>();
-  };
-  
-  template<>
-  class command_graph<graph_state::executable>{
-      public:
-      command_graph() = delete;
-  };
-  
-}
+Exceptions:
 
-----
+* TODO - Throw if this introduces a cycle?
```

Comment:
Do we want `make_edge`  to throw an exception if it is used to introduce a cycle to the graph?

Reply:
This could make the function `make_edge` implicitly become a costly operation especially for graphs at large scale. Because you wold have to run the analysis each time you connect two nodes. Another option would be throwing when the executable graph is created. And/or keeping it optional.

Reply:
Having this error when the finalized graph is created instead makes sense to avoid repeatedly paying the cost of analysing the graph. Seems like the kind of check that would be useful for users while developing an application, but once they actually want to release the application would be unwanted overhead.

So if there's a mechanism for making this optional, that could be good. CUDA has [cudaGraphInstantiateWithFlags](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1ga2c652a24ba93e52b99a47bec0888233),
we could potentially add a flags parameter to `finalize()` that would allow a user to pass in configuration information for creating the executable graph? E.g we could have a `validate` flag that would do this cycle check and possibly other verifications. Other optimization flags could be added here in the future if deemed necessary.

Reply:
Moved this discussion to https://github.com/reble/llvm/issues/12, and will remove this TODO